src/test/groovy/pl/allegro/tech/discovery/consul/recipes/watch/ConsulWatcherTest.groovy

```diff
@@ -267,4 +267,28 @@ class ConsulWatcherTest extends Specification {
         where:
         errorCode << [403, 500]
     }
+
+    def "should not invoke callbacks for cancelled requests"() {
+        given:
+        def callCount = 0
+
+        consul.stubFor(get(urlPathEqualTo('/cancel'))
+                .willReturn(aResponse()
+                .withHeader('X-Consul-Index', '123')
+                .withBody('123')))
+
+        when:
+        watcher
+                .watchEndpoint("/cancel", { it -> callCount += 1}, { it -> callCount += 1})
+                .dispose()
```

Comment:
Isn't it a potential flaky test? What if the call will be so quick that it executes before `.dispose()`. Is this possible?

Also `should not invoke callbacks for cancelled requests`, request are cancelled when you close watcher. Maybe it should be `should not invoke callbacks for disposed watcher`?

Reply:
The name is changed, and I added a fixed delay to the response, to prevent any potential flakiness

Reply:
In the issue you write: 
```
When 2 threads start watches for the same endpoint, and one decides to cancel the "subscription" by running `.stopWatchingEndpoint(String Endpoint)` it results in canceling both subscriptions.

Or it introduces a race condition that results in leaving both subscriptions running.
```

So maybe we should have a second watcher for the same endpoint that we verify still works after cancelling the first one?

Reply:
second subscription added